### PR TITLE
[LI-HOTFIX] Fix flakiness in ControllerIntegrationTest#testRollingControlledShutdown

### PR DESCRIPTION
**Pull request by lmr3796**
This test is introduced in #187 

In the test, it verifies through `getPartitionInfo(topic,partition).get.isr.size`.
However, the partition info may not be ready yet and the .get may throw
exception occasionally.

TICKET = N/A
EXIT_CRITERIA = N/A

### CODE REVIEW

**Comment by lmr3796** [comment]
The test belongs to `intTest/core (A B C)`

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala [hunk 1]
```diff
@@ -667,15 +667,24 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
...
+      )
+    }
+
     // Attempt to shut down and then restart broker1
     val broker1 = servers.filter(s => s.config.brokerId == 1).head
     broker1.shutdown()
     var activeServers = servers.filter(s => s.config.brokerId != 1)
+    waitUntilPartitionInfoAvailableInBrokers(topic, partition, activeServers, "Topic test not created or metadata not propagated after timeout")
     TestUtils.waitUntilTrue(() =>
       activeServers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.isr.size != 2),
```
Comment (gitlw) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Thanks for the fix. Can you please add the isDefined check into this line?

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala [hunk 2]
```diff
@@ -667,15 +667,24 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
...
       "Topic test not created after timeout")
 
     broker1.startup()
+    waitUntilPartitionInfoAvailableInBrokers(topic, partition, servers, "Topic test not created or metadata not propagated after timeout")
     TestUtils.waitUntilTrue(() =>
       servers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic, partition).get.isr.size == 2),
```
Comment (gitlw) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: ditto.

Comment (lmr3796) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: It is actually checked in the patch.  I wrote a function waitUntilPartitionInfoAvailableInBrokers  that checks and performed it in both places

Comment (lmr3796) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Got your point.  Merged the loops

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala [hunk 3]
```diff
@@ -667,26 +667,42 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
...
+    waitUntilIsrSizeConditionHolds(topic, partition, activeServers,
+                                   isrSizePredicate = notEqualsTo(2),
+                                   "Topic test not created or metadata not propagated after timeout")
 
     broker1.startup()
-    TestUtils.waitUntilTrue(() =>
-      servers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic, partition).get.isr.size == 2),
-      "The ISR does not include the full set after the offline broker is restarted")
+    waitUntilIsrSizeConditionHolds(topic, partition, servers,
+                                   isrSizePredicate = equalsTo(2),
```
Comment (gitlw) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Probably replace 
isrSizePredicate = equalsTo(2)
with
_ == 2
considering the latter is shorter and seems easier for understanding.

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala [hunk 4]
```diff
@@ -667,26 +667,42 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
...
 
     // now try to shutdown broker 0
     val broker0 = servers.filter(s => s.config.brokerId == 0).head
     broker0.shutdown()
     activeServers = Seq(broker1)
-    TestUtils.waitUntilTrue(() =>
-      activeServers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.isr.size != 2),
-      "Topic test not created after timeout")
+    waitUntilIsrSizeConditionHolds(topic, partition, activeServers,
+                                   isrSizePredicate = notEqualsTo(2),
```
Comment (gitlw) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: ditto.

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala (hunk 3 above)
Comment (lmr3796) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Fixed

Comment (lmr3796) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Fixed

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala (hunk 4 above)
Comment (lmr3796) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Fixed

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala (hunk 1 above)
Comment (lmr3796) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Fixed

**Review by gitlw — APPROVED** [review]
Thanks for the fix.

**Pull request merged**